### ISSUE TITLE
undocumented "hanging" behavior if `getUnusedAppRestrictionsStatus.get` is called twice in the same thread

### ISSUE THREAD

**Issue by shankari** [e-mission/e-mission-docs#710]
Since yesterday afternoon, CanBikeCO has been hanging intermittently for me on my personal phone.
I get this message periodically:

I first noticed this at around 5pm yesterday (Monday).
I killed the app at around 5:08 using the prompt, but it still hung on relaunch
Until it started working again around 5:15.

Also happened today morning at around 8:44am.
This time I killed the app using the "swipe up" and it still hung until it automatically resolved by around 10am.

Also, the foreground service was killed at least once last night (Monday night) and only restarted after "had to look in all".
And today morning, the app had apparently "stopped tracking", but on launch could not turn on tracking because the UI was not responsive. Then there was an error starting the FSM (can't start tracking, please see next prompt, except that there was no next prompt).
Finally everything went back to normal around 10pm

| app has hung | had to look in all | stopped tracking |
|------|------|-----|
![app_has_hung](https://user-images.githubusercontent.com/2423263/159576254-bb1364bc-abfb-401b-a4ba-e7a0e1cc4924.png) | ![had_to_look_in_all](https://user-images.githubusercontent.com/2423263/159576524-49cec375-41c1-45d4-b563-3f369ced5563.png) | ![stopped_tracking](https://user-images.githubusercontent.com/2423263/159576526-c6acb455-cbb2-4aaf-a0ad-7e3b028675da.png) |

**Comment by shankari** [comment]
Note also that "uploading logs" also hung/I was never able to upload logs directly from the app.
I could only get system logs.

**Comment by shankari** [comment]
All the providers are inaccessible

```
PROVIDER ContentProviderRecord{b2c9b7c u0 gov.colorado.energyoffice.emission/com.google.firebase.provider.FirebaseInitProvider} pid=11640
    Client:
      Failure while dumping the provider: java.io.IOException: Timeout

PROVIDER ContentProviderRecord{91a09de u0 gov.colorado.energyoffice.emission/de.appplant.cordova.emailcomposer.Provider} pid=11640
    Client:
      Failure while dumping the provider: java.io.IOException: Timeout

PROVIDER ContentProviderRecord{411c780 u0 gov.colorado.energyoffice.emission/de.appplant.cordova.plugin.notification.util.AssetProvider} pid=11640
    Client:
      Failure while dumping the provider: java.io.IOException: Timeout

PROVIDER ContentProviderRecord{a802ae1 u0 gov.colorado.energyoffice.emission/nl.xservices.plugins.FileProvider} pid=11640
    Client:
      Failure while dumping the provider: java.io.IOException: Timeout
```

**Comment by shankari** [comment]
here's what I see in the logs.

#### Could not handle the trip end transition since app was in the background. Probably because foreground service was killed. Need to look at last night's logs for more details about when service was killed.

```
03-21 21:14:05.268 10279 18856 18856 E AndroidRuntime: Process: gov.colorado.energyoffice.emission, PID: 18856
03-21 21:14:05.268 10279 18856 18856 E AndroidRuntime: java.lang.RuntimeException: Unable to start receiver edu.berkeley.eecs.emission.cordova.tracker.location.TripDiaryStateMachineReceiver: java.lang.IllegalStateException: Not allowed to start service Intent { act=local.transition.stopped_moving cmp=gov.colorado.energyoffice.emission/edu.berkeley.eecs.emission.cordova.tracker.location.TripDiaryStateMachineService }: app is in background uid UidRecord{d70f8d6 u0a279 RCVR idle change:uncached procs:1 seq(0,0,0)}
03-21 21:14:05.268 10279 18856 18856 E AndroidRuntime:  at android.app.ActivityThread.handleReceiver(ActivityThread.java:3990)
03-21 21:14:05.268 10279 18856 18856 E AndroidRuntime:  at android.app.ActivityThread.access$1500(ActivityThread.java:233)
03-21 21:14:05.268 10279 18856 18856 E AndroidRuntime:  at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2006)
03-21 21:14:05.268 10279 18856 18856 E AndroidRuntime:  at android.os.Handler.dispatchMessage(Handler.java:107)
03-21 21:14:05.268 10279 18856 18856 E AndroidRuntime:  at android.os.Looper.loop(Looper.java:241)
03-21 21:14:05.268 10279 18856 18856 E AndroidRuntime:  at android.app.ActivityThread.main(ActivityThread.java:7582)
03-21 21:14:05.268 10279 18856 18856 E AndroidRuntime:  at java.lang.reflect.Method.invoke(Native Method)
03-21 21:14:05.268 10279 18856 18856 E AndroidRuntime:  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
03-21 21:14:05.268 10279 18856 18856 E AndroidRuntime:  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:941)
03-21 21:14:05.268 10279 18856 18856 E AndroidRuntime: Caused by: java.lang.IllegalStateException: Not allowed to start service Intent { act=local.transition.stopped_moving cmp=gov.colorado.energyoffice.emission/edu.berkeley.eecs.emission.cordova.tracker.location.TripDiaryStateMachineService }: app is in background uid UidRecord{d70f8d6 u0a279 RCVR idle change:uncached procs:1 seq(0,0,0)}
03-21 21:14:05.268 10279 18856 18856 E AndroidRuntime:  at android.app.ContextImpl.startServiceCommon(ContextImpl.java:1637)
03-21 21:14:05.268 10279 18856 18856 E AndroidRuntime:  at android.app.ContextImpl.startService(ContextImpl.java:1592)
03-21 21:14:05.268 10279 18856 18856 E AndroidRuntime:  at android.content.ContextWrapper.startService(ContextWrapper.java:669)
03-21 21:14:05.268 10279 18856 18856 E AndroidRuntime:  at android.content.ContextWrapper.startService(ContextWrapper.java:669)
03-21 21:14:05.268 10279 18856 18856 E AndroidRuntime:  at edu.berkeley.eecs.emission.cordova.tracker.location.TripDiaryStateMachineReceiver.onReceive(TripDiaryStateMachineReceiver.java:117)
03-21 21:14:05.268 10279 18856 18856 E AndroidRuntime:  at android.app.ActivityThread.handleReceiver(ActivityThread.java:3981)
03-21 21:14:05.268 10279 18856 18856 E AndroidRuntime:  ... 8 more
03-21 21:54:44.044  root   632   632 F libc    : crash_dump helper failed to exec
03-21 21:55:14.521  root   632   632 F libc    : crash_dump helper failed to exec
03-21 22:02:48.056  root   632   632 F libc    : crash_dump helper failed to exec
03-21 22:15:16.041  root   632   632 F libc    : crash_dump helper failed to exec
03-21 22:17:27.888  root   632   632 F libc    : crash_dump helper failed to exec
```

**Comment by shankari** [comment]
Ok so here is an app launch from today evening; timeouts up the wazoo...

```
03-22 18:33:52.634  1000  1669 10695 I ActivityTaskManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=gov.colorado.energyoffice.emission/.MainActivity bnds=[293,68][427,307]} from uid 10193
03-22 18:33:52.720  1000  1669  1702 I WindowManager:   SURFACE show Surface(name=Splash Screen gov.colorado.energyoffice.emission)/@0x66f2e60 on display:0: Splash Screen gov.colorado.energyoffice.emission
03-22 18:34:36.813  1000  1669  1701 W ActivityTaskManager: Activity top resumed state loss timeout for ActivityRecord{b86a3d7 u0 gov.colorado.energyoffice.emission/.MainActivity
 t1686}
03-22 18:34:36.821  1000  1669  1701 W ActivityTaskManager: Activity pause timeout for ActivityRecord{b86a3d7 u0 gov.colorado.energyoffice.emission/.MainActivity t1686}
03-22 18:34:47.839  1000  1669  1701 W ActivityTaskManager: Activity stop timeout for ActivityRecord{b86a3d7 u0 gov.colorado.energyoffice.emission/.MainActivity t1686}
03-22 18:34:47.839  1000  1669  1701 I ActivityTaskManager: Activity reported stop, but no longer stopping: ActivityRecord{b86a3d7 u0 gov.colorado.energyoffice.emission/.MainActivity t1686}
03-22 18:34:56.444  1000  1669  1701 W ActivityTaskManager: Activity top resumed state loss timeout for ActivityRecord{b86a3d7 u0 gov.colorado.energyoffice.emission/.MainActivity t1686}
03-22 18:34:56.445  1000  1669  1701 W ActivityTaskManager: Activity pause timeout for ActivityRecord{b86a3d7 u0 gov.colorado.energyoffice.emission/.MainActivity t1686}
03-22 18:35:07.452  1000  1669  1701 W ActivityTaskManager: Activity stop timeout for ActivityRecord{b86a3d7 u0 gov.colorado.energyoffice.emission/.MainActivity t1686}
03-22 18:37:21.065  1000  1669  1701 W ActivityTaskManager: Activity stop timeout for ActivityRecord{b86a3d7 u0 gov.colorado.energyoffice.emission/.MainActivity t1686}
```

I don't know what the `am_*` logs are about, but it looks like there was a ANR in the sync and it was killed.

```
03-22 17:59:28.767  1000  1669  1709 I am_anr  : [0,11597,gov.colorado.energyoffice.emission:sync,818462276,executing service gov.colorado.energyoffice.emission/edu.berkeley.eecs.emission.cordova.serversync.SyncService]
03-22 17:59:28.946  1000  1669  1709 I am_kill : [0,11597,gov.colorado.energyoffice.emission:sync,0,bg anr]
03-22 17:59:28.966  1000  1669 10342 I am_proc_died: [0,11597,gov.colorado.energyoffice.emission:sync,0,11]
```

and then we see

```
03-22 18:33:52.660  1000  1669 10695 I am_create_activity: [0,193373143,1686,gov.colorado.energyoffice.emission/.MainActivity,android.intent.action.MAIN,NULL,NULL,270532608]
03-22 18:33:52.674  1000  1669  8241 I am_restart_activity: [0,193373143,1686,gov.colorado.energyoffice.emission/.MainActivity]
03-22 18:33:52.675  1000  1669  8241 I am_set_resumed_activity: [0,gov.colorado.energyoffice.emission/.MainActivity,minimalResumeActivityLocked]
03-22 18:34:02.707  1000  1669  1708 I am_pss  : [11640,10279,gov.colorado.energyoffice.emission,7805952,757760,5550080,51994624,0,2,7]
03-22 18:34:03.293  1000  1669  1708 I am_pss  : [14780,10231,com.google.android.apps.docs,50863104,29589504,16263168,102170624,0,20,14]
03-22 18:34:18.588  1000  1669  1708 I am_pss  : [11640,10279,gov.colorado.energyoffice.emission,7805952,757760,5550080,51994624,0,2,7]
03-22 18:34:36.314  1000  1669  1691 I am_pause_activity: [0,193373143,gov.colorado.energyoffice.emission/.MainActivity,userLeaving=true]
03-22 18:34:36.316  1000  1669  1691 I am_add_to_stopping: [0,193373143,gov.colorado.energyoffice.emission/.MainActivity,makeInvisible]
03-22 18:34:36.837  1000  1669  1701 I am_stop_activity: [0,193373143,gov.colorado.energyoffice.emission/.MainActivity]
```

That seems fine, and I don't actually seen any other ANRs in this log.

**Comment by shankari** [comment]
Poking around the dump, found a bunch of ANRs, all related to the CEO sync.

```
FS/data/anr//anr_2022-03-22-17-38-50-279:Cmd line: gov.colorado.energyoffice.emission:sync
FS/data/anr//anr_2022-03-22-17-38-50-279:#2 dalvik.system.PathClassLoader: [/data/app/gov.colorado.energyoffice.emission-qHIIc6xMVGB8rWNLakZlIA==/base.apk], parent #1
FS/data/anr//anr_2022-03-22-17-38-50-279:/data/app/gov.colorado.energyoffice.emission-qHIIc6xMVGB8rWNLakZlIA==/oat/arm64/base.odex: speed-profile
```

Looking at it in detail, we get that the main thread is blocked in

```
"main" prio=5 tid=1 Waiting
  | group="main" sCount=1 dsCount=0 flags=1 obj=0x71847fe8 self=0x782829dc00
  | sysTid=29453 nice=0 cgrp=default sched=0/0 handle=0x7829803ed0
  | state=S schedstat=( 58440316 4705989 106 ) utm=2 stm=3 core=2 HZ=100
  | stack=0x7fc2630000-0x7fc2632000 stackSize=8192KB
  | held mutexes=
  at sun.misc.Unsafe.park(Native method)
  - waiting on an unknown object
  at java.util.concurrent.locks.LockSupport.park(LockSupport.java:190)
  at androidx.concurrent.futures.AbstractResolvableFuture.get(AbstractResolvableFuture.java:485)
  at edu.berkeley.eecs.emission.cordova.tracker.verification.SensorControlChecks.checkUnusedAppsUnrestricted(SensorControlChecks.java:108)
  at edu.berkeley.eecs.emission.cordova.tracker.verification.SensorControlBackgroundChecker.lambda$checkAppState$0(SensorControlBackgroundChecker.java:96)
  at edu.berkeley.eecs.emission.cordova.tracker.verification.-$$Lambda$SensorControlBackgroundChecker$7ewMbTXs4b1E0GgjdjTWryZ_onw.onComplete(lambda:-1)
  at com.google.android.gms.tasks.zzj.run(com.google.android.gms:play-services-tasks@@17.2.0:4)
  - locked <0x04f8118e> (a java.lang.Object)
  at android.os.Handler.handleCallback(Handler.java:914)
  at android.os.Handler.dispatchMessage(Handler.java:100)
  at android.os.Looper.loop(Looper.java:241)
  at android.app.ActivityThread.main(ActivityThread.java:7582)
  at java.lang.reflect.Method.invoke(Native method)
```

**Comment by shankari** [comment]
All the ANRs seem to be related

```
system_logs_tue_evening_hung$ ls -1 FS/data/anr/* | wc -l
      10
system_logs_tue_evening_hung$ grep checkUnusedAppsUnrestricted FS/data/anr/* | wc -l
      10
```

and 

```
$ grep checkUnusedAppsUnrestricted system_logs_mon_evening/FS/data/anr/anr_2022-03-21-17-* | wc -l
      10
$ ls -1 system_logs_mon_evening/FS/data/anr/anr_2022-03-21-17-* | wc -l
      10
```

**Comment by shankari** [comment]
So it looks like `checkUnusedAppsUnrestricted` is hanging for some reason. I do note that I now have a message saying that unused app restrictions will now be removed. Looking at the notification timestamp, this was ~ 17 h before 9:48pm, so around Monday morning.

The google blog
https://android-developers.googleblog.com/2021/09/making-permissions-auto-reset-available.html

says that:
> Q1 2022 - The permission auto-reset feature will reach all devices running a version between Android 6.0 and Android 10. 

So it looks like the feature reached my device, and then the calls to `getUnusedAppRestrictionsStatus` started hanging.
`getUnusedAppRestrictionsStatus` returns a `ListenableFuture`, so a fairly simple fix is to get the result asynchronously using a listener instead of synchronously using `future.get()`.

```
      ListenableFuture<Integer> future = PackageManagerCompat.getUnusedAppRestrictionsStatus(ctxt);
    try {
      Integer appRestrictionStatus = future.get();
```

Would be good to reproduce this in the emulator if possible and test the fix before making it randomly,

**Comment by shankari** [comment]
The Google Play services version on my phone is version 22.06.18 (100400-433619428)
The Google Play services version on M's phone (which does not have this behavior) is 20.39.15 (040306-335085812)

**Comment by shankari** [comment]
Ok on an android API 29 emulator, version 17.7.86, calling `checkUnusedAppsUnrestricted` calls `future.get()`, which instantly returns 

```
future = {ResolvableFuture@11758} "androidx.concurrent.futures.ResolvableFuture@12cc9c2[status=SUCCESS, result=[1]]"
 listeners = {AbstractResolvableFuture$Listener@11762} 
 value = {Integer@11744} 1
 waiters = {AbstractResolvableFuture$Waiter@11763} 
 shadow$_klass_ = {Class@9451} "class androidx.concurrent.futures.ResolvableFuture"
 shadow$_monitor_ = -2127771198
```

where 1 = `UnusedAppRestrictionsConstants.FEATURE_NOT_AVAILABLE`

**Comment by shankari** [comment]
Couldn't find google play services in the list of apps in the play store, so upgraded google, which prompted an upgrade to google play services. It is now 15.2.67. Still not recent enough. Installed downloaded apk from softopedia *in the emulator*.
It is now 22.06.57 but is restarting like crazy.
`get` works the same way.

**Comment by shankari** [comment]
Ok, so then figured out how to launch the play store entry for it:
- use URL https://play.google.com/store/apps/details?id=com.google.android.gms
- click on "App installed from Google Play"

After clicking around a bit, it is now at 22.06.18. Woo!

**Comment by shankari** [comment]
Maybe this the play store instead of play services.
Again, play store doesn't show up as an app in itself, but following this instructions (https://support.google.com/googleplay/thread/108116960/how-to-update-play-store?hl=en), my phone is at 29.7.17, the emulator is at 22.4.25. Upgrading it...
Had to manually download and install the apk, but no crashes and the play store UI is the same as mine.

**Comment by shankari** [comment]
Ok, so here's what I know so far:
- the unused app restrictions were backported into google play store version 29.7.17
- before that update, `getUnusedAppRestrictionsStatus` returned `FEATURE_NOT_AVAILABLE`
- after that update, `getUnusedAppRestrictionsStatus` returns ``API_30_BACKPORT`

There is no performance difference in the emulator.
However, there does appear to be a performance difference in the physical phone.
This kind of makes sense, because presumably `API_30_BACKPORT` has to perform some more complicated checks than just returning `FEATURE_NOT_AVAILABLE`
Not sure why the performance difference is not visible in the emulator but could be because of additional resources or the fact that it is an emulator so cuts some corners in the computation.

The fix is probably (as discussed) to switch to async operation.
- we can do this either at the `checkUnusedAppsUnrestricted` level by adding a listener, or
- in the callback function as recommended by cordova, or
- both

**Comment by shankari** [comment]
`checkUnusedAppsUnrestricted` is called from both the foreground and the background. Making the callback function be asynchronous only fixes the foreground case. Not quite sure what happens in the background case, but maybe that is the reason that we had that issue with starting today morning.

Let's start with fixing `checkUnusedAppsUnrestricted`. Probably easiest to return a task, similar to the existing `checkLocationSettings`. This will allow us to wait for both tasks. But is that even the right thing to do if the task is hanging...

**Comment by shankari** [comment]
Note that the current hang is happening from the background checker code while we tested in the emulator from the foreground check code.

```
  at androidx.concurrent.futures.AbstractResolvableFuture.get(AbstractResolvableFuture.java:485)
  at edu.berkeley.eecs.emission.cordova.tracker.verification.SensorControlChecks.checkUnusedAppsUnrestricted(SensorControlChecks.java:108)
  at edu.berkeley.eecs.emission.cordova.tracker.verification.SensorControlBackgroundChecker.lambda$checkAppState$0(SensorControlBackgroundChecker.java:96)
  at edu.berkeley.eecs.emission.cordova.tracker.verification.-$$Lambda$SensorControlBackgroundChecker$7ewMbTXs4b1E0GgjdjTWryZ_onw.onComplete(lambda:-1)
  at com.google.android.gms.tasks.zzj.run(com.google.android.gms:play-services-tasks@@17.2.0:4)
  - locked <0x04f8118e> (a java.lang.Object)
```

**Comment by shankari** [comment]
Aha! Can reproduce it now! Even from the foreground. As far as I can remember, steps to reproduce are:
- Launch app
- Fix background restriction settings (opens play store)
- can't figure out a way to get back
- launch app from home screen
- hangs

**Comment by shankari** [comment]
Seems to be 100% reproducible:
- Doesn't hang on first launch

```
2022-03-22 23:26:10.413 27401-27800/edu.berkeley.eecs.emission I/SensorControlChecks: About to call future.get to read the restriction status
2022-03-22 23:26:10.486 27401-27800/edu.berkeley.eecs.emission I/SensorControlChecks: Received 3 from future.get
```

- refresh background restrictions, still works

```
2022-03-22 23:28:16.398 27401-27800/edu.berkeley.eecs.emission I/SensorControlChecks: About to call future.get to read the restriction status
2022-03-22 23:28:16.427 27401-27800/edu.berkeley.eecs.emission I/SensorControlChecks: Received 3 from future.get
```

- fix background restrictions; couldn't find app main screen after fix, went back/back/back...

```
2022-03-22 23:29:41.456 27401-27401/edu.berkeley.eecs.emission I/SensorControlChecks: About to call future.get to read the restriction status
```

- now hung and not proceeding. not sure what it would have returned since the call still hasn't even returned.

so this is not just an issue from the background, but the app must just be calling this function from the background periodically and ANRing every time.

**Comment by shankari** [comment]
On, so after kill + restart, the call succeeds at first

```
2022-03-22 23:50:48.263 32108-32244/edu.berkeley.eecs.emission I/SensorControlChecks: About to call future.get to read the restriction status
2022-03-22 23:50:48.278 32108-32244/edu.berkeley.eecs.emission I/SensorControlChecks: Received 2 from future.get
2022-03-22 23:50:59.253 32108-32244/edu.berkeley.eecs.emission I/SensorControlChecks: About to call future.get to read the restriction status
2022-03-22 23:50:59.271 32108-32244/edu.berkeley.eecs.emission I/SensorControlChecks: Received 2 from future.get
2022-03-22 23:51:03.310 32108-32244/edu.berkeley.eecs.emission I/SensorControlChecks: About to call future.get to read the restriction status
2022-03-22 23:51:03.328 32108-32244/edu.berkeley.eecs.emission I/SensorControlChecks: Received 2 from future.get
```

Then, the background service gets a couple more initialize calls. One of which succeeds.

```
2022-03-22 23:51:03.344 32108-32244/edu.berkeley.eecs.emission I/SensorControlBackgroundChecker: in restartFSMIfStartState, currState = local.state.start
2022-03-22 23:51:03.356 32108-32244/edu.berkeley.eecs.emission I/SensorControlBackgroundChecker: in start state, sending initialize
2022-03-22 23:51:35.981 32108-32108/edu.berkeley.eecs.emission I/SensorControlBackgroundChecker: in restartFSMIfStartState, currState = local.state.start
2022-03-22 23:51:35.996 32108-32108/edu.berkeley.eecs.emission I/SensorControlBackgroundChecker: in start state, sending initialize
2022-03-22 23:51:36.104 32108-32244/edu.berkeley.eecs.emission I/SensorControlChecks: About to check location settings
2022-03-22 23:51:36.152 32108-32244/edu.berkeley.eecs.emission D/SensorControlChecks: Checking location settings for request Request[PRIORITY_HIGH_ACCURACY requested=30000ms fastest=5000ms]
2022-03-22 23:51:36.167 32108-32244/edu.berkeley.eecs.emission D/SensorControlChecks: Got back result com.google.android.gms.tasks.zzu@9fcb158
2022-03-22 23:51:36.278 32108-32244/edu.berkeley.eecs.emission I/SensorControlChecks: About to call future.get to read the restriction status
2022-03-22 23:51:36.300 32108-32244/edu.berkeley.eecs.emission I/SensorControlChecks: Received 2 from future.get
```

And then there is a third initialize, which hangs

```
2022-03-22 23:51:41.203 32108-32108/edu.berkeley.eecs.emission I/SensorControlBackgroundChecker: in restartFSMIfStartState, currState = local.state.start
2022-03-22 23:51:41.216 32108-32108/edu.berkeley.eecs.emission I/SensorControlBackgroundChecker: in start state, sending initialize
```

The second call from before succeeds.

```
2022-03-22 23:51:41.300 32108-32244/edu.berkeley.eecs.emission I/SensorControlChecks: About to check location settings
2022-03-22 23:51:41.346 32108-32244/edu.berkeley.eecs.emission D/SensorControlChecks: Checking location settings for request Request[PRIORITY_HIGH_ACCURACY requested=30000ms fastest=5000ms]
2022-03-22 23:51:41.357 32108-32244/edu.berkeley.eecs.emission D/SensorControlChecks: Got back result com.google.android.gms.tasks.zzu@fd3af6e
2022-03-22 23:51:41.480 32108-32244/edu.berkeley.eecs.emission I/SensorControlChecks: About to call future.get to read the restriction status
2022-03-22 23:51:41.494 32108-32244/edu.berkeley.eecs.emission I/SensorControlChecks: Received 2 from future.get
```

But the third call fails

```
2022-03-22 23:51:44.539 32108-32108/edu.berkeley.eecs.emission I/SensorControlChecks: About to check location settings
2022-03-22 23:51:44.563 32108-32108/edu.berkeley.eecs.emission D/SensorControlChecks: Checking location settings for request Request[PRIORITY_HIGH_ACCURACY requested=30000ms fastest=5000ms]
2022-03-22 23:51:44.572 32108-32108/edu.berkeley.eecs.emission D/SensorControlChecks: Got back result com.google.android.gms.tasks.zzu@27cb593
2022-03-22 23:51:44.583 32108-32108/edu.berkeley.eecs.emission I/SensorControlBackgroundChecker: All settings are valid, checking current state
2022-03-22 23:51:44.591 32108-32108/edu.berkeley.eecs.emission I/SensorControlBackgroundChecker: Current location settings are com.google.android.gms.location.LocationSettingsResponse@74f7cce
2022-03-22 23:51:44.612 32108-32108/edu.berkeley.eecs.emission I/SensorControlChecks: About to call future.get to read the restriction status
```

Why does the third call fail? What is different between it and the first two?

**Comment by shankari** [comment]
Launching directly from the phone again. Works first  (from foreground).

```
2022-03-22 23:56:50.806 802-948/edu.berkeley.eecs.emission I/SensorControlChecks: About to check location settings
2022-03-22 23:56:50.862 802-948/edu.berkeley.eecs.emission D/SensorControlChecks: Checking location settings for request Request[PRIORITY_HIGH_ACCURACY requested=30000ms fastest=5000ms]
2022-03-22 23:56:50.918 802-948/edu.berkeley.eecs.emission D/SensorControlChecks: Got back result com.google.android.gms.tasks.zzu@4dac46b
2022-03-22 23:56:51.062 802-948/edu.berkeley.eecs.emission I/SensorControlChecks: About to call future.get to read the restriction status
2022-03-22 23:56:51.083 802-948/edu.berkeley.eecs.emission I/SensorControlChecks: Received 2 from future.get
```

Does not work from background.

```
2022-03-22 23:57:05.995 802-802/edu.berkeley.eecs.emission I/SensorControlChecks: About to check location settings
2022-03-22 23:57:06.031 802-802/edu.berkeley.eecs.emission D/SensorControlChecks: Checking location settings for request Request[PRIORITY_HIGH_ACCURACY requested=30000ms fastest=5000ms]
2022-03-22 23:57:06.046 802-802/edu.berkeley.eecs.emission D/SensorControlChecks: Got back result com.google.android.gms.tasks.zzu@6245370
2022-03-22 23:57:06.077 802-802/edu.berkeley.eecs.emission I/SensorControlBackgroundChecker: All settings are valid, checking current state
2022-03-22 23:57:06.097 802-802/edu.berkeley.eecs.emission I/SensorControlBackgroundChecker: Current location settings are com.google.android.gms.location.LocationSettingsResponse@fd3af6e
2022-03-22 23:57:06.134 802-802/edu.berkeley.eecs.emission I/SensorControlChecks: About to call future.get to read the restriction status
```

**Comment by shankari** [comment]
The problem is really from the background call; is this an issue with calling the function multiple times in the same thread or something.

**Comment by shankari** [comment]
Yes, this does seem to be because the function does not appear to be idempotent. It crashes from the foreground if it is called twice in the same thread - e.g. 

```
2022-03-23 00:49:58.236 21001-21558/edu.berkeley.eecs.emission I/SensorControlChecks: About to call future.get to read the restriction status
2022-03-23 00:49:58.263 21001-21558/edu.berkeley.eecs.emission I/SensorControlChecks: Received 3 from future.get
2022-03-23 00:49:58.284 21001-21558/edu.berkeley.eecs.emission I/SensorPermissionsAndSettingsForegroundDelegate: Unused apps restricted, asking user to unrestrict
2022-03-23 00:49:59.356 21001-21001/edu.berkeley.eecs.emission D/CordovaActivity: Stopped the activity.
0 (tinfo 0xe430f980)
2022-03-23 00:50:05.723 21001-21001/edu.berkeley.eecs.emission D/CordovaActivity: Incoming Result. Request code = 362253743
2022-03-23 00:50:05.723 21001-21001/edu.berkeley.eecs.emission D/CordovaInterfaceImpl: Sending activity result to plugin
2022-03-23 00:50:05.744 21001-21001/edu.berkeley.eecs.emission D/DataCollectionPlugin: received onActivityResult(362253743,-1,null data)
2022-03-23 00:50:05.769 21001-21001/edu.berkeley.eecs.emission D/SensorPermissionsAndSettingsForegroundDelegate: 362253743 is our code, handling callback
2022-03-23 00:50:05.790 21001-21001/edu.berkeley.eecs.emission D/SensorPermissionsAndSettingsForegroundDelegate: Got unused app restrictions callback from launching app settings
2022-03-23 00:50:05.819 21001-21001/edu.berkeley.eecs.emission I/SensorControlChecks: About to call future.get to read the restriction status
```

Basically, we should ensure that all calls to `checkUnusedAppsUnrestricted` that are not a direct UI call should be called asynchronously. And we might want to extend that to the UI calls as well.

Used in:
- `SensorControlBackgroundChecker`: invoked within `AsyncTask`
- `SensorControlForegroundChecker`:
   - `checkUnusedAppsUnrestricted`: direct UI call
   - `checkAndPromptUnusedAppsUnrestricted`: direct UI call
   - `onActivityResult`: invoked within `AsyncTask`

**Comment by shankari** [comment]
Hm, the fix is not that simple, using an AsyncTask seems to leak the connection to the backport service.

```
2022-03-23 00:59:02.607 29247-29247/edu.berkeley.eecs.emission E/ActivityThread: Service edu.berkeley.eecs.emission.cordova.tracker.location.TripDiaryStateMachineService has leaked ServiceConnection androidx.core.content.UnusedAppRestrictionsBackportServiceConnection@12e3e46 that was originally bound here
    android.app.ServiceConnectionLeaked: Service edu.berkeley.eecs.emission.cordova.tracker.location.TripDiaryStateMachineService has leaked ServiceConnection androidx.core.content.UnusedAppRestrictionsBackportServiceConnection@12e3e46 that was originally bound here
        at android.app.LoadedApk$ServiceDispatcher.<init>(LoadedApk.java:1805)
        at android.app.LoadedApk.getServiceDispatcherCommon(LoadedApk.java:1677)
        at android.app.LoadedApk.getServiceDispatcher(LoadedApk.java:1656)
        at android.app.ContextImpl.bindServiceCommon(ContextImpl.java:1721)
        at android.app.ContextImpl.bindService(ContextImpl.java:1650)
        at android.content.ContextWrapper.bindService(ContextWrapper.java:705)
        at androidx.core.content.UnusedAppRestrictionsBackportServiceConnection.connectAndFetchResult(UnusedAppRestrictionsBackportServiceConnection.java:110)
        at androidx.core.content.PackageManagerCompat.getUnusedAppRestrictionsStatus(PackageManagerCompat.java:189)
        at edu.berkeley.eecs.emission.cordova.tracker.verification.SensorControlChecks.checkUnusedAppsUnrestricted(SensorControlChecks.java:106)
        at edu.berkeley.eecs.emission.cordova.tracker.verification.SensorControlBackgroundChecker$1.run(SensorControlBackgroundChecker.java:133)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:289)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:919)
2022-03-23 00:59:02.613 29247-29514/edu.berkeley.eecs.emission E/AndroidRuntime: FATAL EXCEPTION: pool-7-thread-1
    Process: edu.berkeley.eecs.emission, PID: 29247
    java.lang.IllegalArgumentException: Service not registered: androidx.core.content.UnusedAppRestrictionsBackportServiceConnection@12e3e46
        at android.app.LoadedApk.forgetServiceDispatcher(LoadedApk.java:1751)
        at android.app.ContextImpl.unbindService(ContextImpl.java:1776)
        at android.content.ContextWrapper.unbindService(ContextWrapper.java:741)
        at androidx.core.content.UnusedAppRestrictionsBackportServiceConnection.disconnectFromService(UnusedAppRestrictionsBackportServiceConnection.java:118)
        at androidx.core.content.-$$Lambda$SJcDCtX_JtAI8Mo8DbauUqPURzc.run(Unknown Source:2)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:919)
2022-03-23 00:59:16.657 29247-29292/edu.berkeley.eecs.emission E/FirebaseInstanceId: Token retrieval failed: SERVICE_NOT_AVAILABLE
2022-03-23 01:00:16.679 29247-29292/edu.berkeley.eecs.emission E/FirebaseInstanceId: Token retrieval failed: SERVICE_NOT_AVAILABLE
```

**Comment by shankari** [comment]
Once the install is complete, the background check works fine with the AsyncTask. The problem appears to be only during the initial install

**Comment by shankari** [comment]
This seems to be a timing issue - if I add a breakpoint in `checkUnusedAppsUnrestricted` and step through slowly, it works. Argh! This is terrible. As we can see from the stacktrace, I don't manage the backported connections, `androidx.core.content.UnusedAppRestrictionsBackportServiceConnection.connectAndFetchResult(UnusedAppRestrictionsBackportServiceConnection.java:110)` does. I simply call `checkUnusedAppsUnrestricted`.

It turns out that the service connection is not only non-reentrant, it also leaks connections based on timing.

And of course, 
https://developer.android.com/reference/androidx/core/content/PackageManagerCompat#getUnusedAppRestrictionsStatus(android.content.Context)
does not document any of this.

I need to file an issue against android.core. But I need to create an MRE for that.
Let's work around this for now.
- Launch the AsyncTask after a minute or so: background restriction background check has no dependencies so this shouldn't hurt anything and might help with the timing issue
- Catch the exception and see if we can close the connection manually.

**Comment by shankari** [comment]
so to summarize, the main problem is that the `getUnusedAppRestrictionsStatus` can apparently not be called twice in the same thread.

We fix this by:
- moving the callback in the ForegroundServiceDelegate into an `AsyncTask`
- moving the call in the BackgroundChecker into an executor

Building this and pushing out to beta right now, hopefully can promote to production before the meeting today afternoon

**Comment by shankari** [comment]
Testing this a few times; still crashes intermittently (around 50% of the time). Again, seems to depend on timing.
Delaying execution by one minute using the `ScheduledExecutorService` seems to work. Have launched the app 5 times, super fast, with no crashes.

**Comment by shankari** [comment]
Dashboard from google play store showing:
- huge increase in ANR from around Mar 14, increasing steadily as the feature rolls out to more phones
- indication that the issue is only with android 10 and below. Android 11 is fine.

<img width="1464" alt="Screen Shot 2022-03-23 at 9 45 05 AM" src="https://user-images.githubusercontent.com/2423263/159751871-30506037-e9e5-46a5-a23c-5cd1dd5901fb.png">
<img width="1464" alt="Screen Shot 2022-03-23 at 9 43 56 AM" src="https://user-images.githubusercontent.com/2423263/159751686-8e2f9b0f-bf50-444d-ae3c-e3d417931754.png">

**Comment by shankari** [comment]
pushed out to "internal testing". Upgraded immediately on my phone.
App launched fine, was able to check dashboard.
Got a notification about incorrect app settings.
Tried to handle it and app status opened with background restrictions section red
killed the app to avoid setting it to green; will take screenshots the next time it happens.

After that, will promote to product to start getting the ANR count down.

**Comment by shankari** [comment]
ah, so got incorrect app settings prompt, but the app hung when I clicked on it.
So I killed by swiping up, and then the app was no longer hanging.

Tried "force sync" a couple of times to try and reproduce and generated a system log dump
Notification showed up after around a minute or so.

Clicking on this notification worked (at around 11:37)

Trying again when the app is not active.
Will probably push right now anyway to get the ANRs under control but will try to test and fix further tonight.

**Comment by shankari** [comment]
11:48: force sync from app; app sent to background
11:49: notification
11:52: kill app by swiping up; notification goes away
11:53: relaunch app; force sync; send app to background
11:54: notification (remember one minute delay!)
12:19: click on notification from locked state - works!

Going to promote this for now while debugging the intermittent hanging (if any) for the rest of the day.

**Comment by shankari** [comment]
pushed out to internal testing at around 10am. Last ANR was at `anr_2022-03-23-09-34-58-422`

Unfortunately, detailed activity manager logs are only from 11:32, and the launch was closer to 11:30.

```
03-23 11:32:45.412  1000  1669  8241 I am_set_resumed_activity: [0,gov.colorado.energyoffice.emission/.MainActivity,bringingFoundTaskToFront]
03-23 11:32:45.413  1000  1669  8241 I am_new_intent: [0,182875583,1712,gov.colorado.energyoffice.emission/.MainActivity,android.intent.action.MAIN,NULL,NULL,270532608]
03-23 11:32:45.430  1000  1669 12503 I am_set_resumed_activity: [0,gov.colorado.energyoffice.emission/.MainActivity,resumeTopActivityInnerLocked]
03-23 11:32:45.432  1000  1669 12503 I am_resume_activity: [0,182875583,1712,gov.colorado.energyoffice.emission/.MainActivity]
```

No ANR in those logs, though

**Comment by shankari** [comment]
1:08: notification
1:09 clicked on it and unlocked. Opened to SMS app (which I had been using before). CanBikeCO showed up as a small bar at the top of the screen (see screenshot)
1:10-ish backgrounded SMS app, notification still in place
1:17: click on notification; same hang; emailed screenshot; bar on top goes away
1:18: launch CanBikeCO manually; launches slowly; notification still present
1:19: click on notification; leave app in foreground for a couple of minutes
1:26: still hung; kill app by swiping up; notification not killed
1:27: click on notification again; app still hung
1:27: kill app again and launch from icon; works
1:28 app in foreground; click notification; again run into weird bar on top
1:50 the weird bar seems to be from a different activity than the main CanBikeCO app - I have the app in the "list of apps" screen and click the notification and it doesn't bring it to the foreground. Instead it adds a new border to it.
    - Is this something weird to do with the ClickActivity and do I have to create the notification from the UI thread after all?
    - Note that when the notification is created from the "force sync" button, it works. It just doesn't work if it is from the background check function.
    - Uploaded logs
    - Generating bug report

**Comment by shankari** [comment]
Bug report still does not show an ANR

```
$ ls -1 FS/data/anr/anr_2022-03-23-0*
FS/data/anr/anr_2022-03-23-00-34-12-796
FS/data/anr/anr_2022-03-23-03-06-53-208
FS/data/anr/anr_2022-03-23-05-53-33-673
FS/data/anr/anr_2022-03-23-06-47-07-860
FS/data/anr/anr_2022-03-23-08-16-15-187
FS/data/anr/anr_2022-03-23-09-34-58-422
```

**Comment by shankari** [comment]
Yup, I was right!

>  1:50 the weird bar seems to be from a different activity than the main CanBikeCO app - I have the app in the "list of apps" screen and click the notification and it doesn't bring it to the foreground. Instead it adds a new border to it.
> - Is this something weird to do with the ClickActivity and do I have to create the notification from the UI thread after all?

around the time that I was fuzting around with clicking on the notification

```
    time="2022-03-23 13:19:39" type=USER_INTERACTION package=gov.colorado.energyoffice.emission flags=0x0
    time="2022-03-23 13:19:39" type=ACTIVITY_PAUSED package=gov.colorado.energyoffice.emission class=gov.colorado.energyoffice.emission.MainActivity instanceId=133170508 taskRootPackage=gov.colorado.energyoffice.emission taskRootClass=gov.colorado.energyoffice.emission.MainActivity flags=0x0
    time="2022-03-23 13:19:39" type=ACTIVITY_RESUMED package=gov.colorado.energyoffice.emission class=de.appplant.cordova.plugin.localnotification.ClickReceiver instanceId=36004953 taskRootPackage=gov.colorado.energyoffice.emission taskRootClass=de.appplant.cordova.plugin.localnotification.ClickReceiver flags=0x0

    time="2022-03-23 13:19:51" type=ACTIVITY_PAUSED package=gov.colorado.energyoffice.emission class=de.appplant.cordova.plugin.localnotification.ClickReceiver instanceId=36004953 taskRootPackage=gov.colorado.energyoffice.emission taskRootClass=de.appplant.cordova.plugin.localnotification.ClickReceiver flags=0x0
    time="2022-03-23 13:19:51" type=ACTIVITY_RESUMED package=com.fsck.k9 class=com.fsck.k9.activity.MessageCompose instanceId=2729269 taskRootPackage=com.fsck.k9 taskRootClass=com.fsck.k9.activity.MessageList flags=0x0
    time="2022-03-23 13:19:52" type=ACTIVITY_STOPPED package=gov.colorado.energyoffice.emission class=de.appplant.cordova.plugin.localnotification.ClickReceiver instanceId=36004953 taskRootPackage=gov.colorado.energyoffice.emission taskRootClass=de.appplant.cordova.plugin.localnotification.ClickReceiver flags=0x0
    time="2022-03-23 13:19:52" type=ACTIVITY_STOPPED package=gov.colorado.energyoffice.emission class=gov.colorado.energyoffice.emission.MainActivity instanceId=133170508 taskRootPackage=gov.colorado.energyoffice.emission taskRootClass=gov.colorado.energyoffice.emission.MainActivity flags=0x0
```

**Comment by shankari** [comment]
Going back to the original click

> 1:08: notification
> 1:09 clicked on it and unlocked. Opened to SMS app (which I had been using before). CanBikeCO showed up as a small bar at the top of the screen (see screenshot)
> 1:10-ish backgrounded SMS app, notification still in place


Notification generated

```
    time="2022-03-23 13:08:45" type=NOTIFICATION_INTERRUPTION package=gov.colorado.energyoffice.emission channelId=default-channel-id flags=0x0
    time="2022-03-23 13:08:45" type=NOTIFICATION_SEEN package=gov.colorado.energyoffice.emission flags=0x0
```

`ClickReceiver` launched

```
    time="2022-03-23 13:09:07" type=USER_INTERACTION package=gov.colorado.energyoffice.emission flags=0x0
    time="2022-03-23 13:09:07" type=ACTIVITY_RESUMED package=gov.colorado.energyoffice.emission class=de.appplant.cordova.plugin.localnotification.ClickReceiver instanceId=101733288 taskRootPackage=gov.colorado.energyoffice.emission taskRootClass=de.appplant.cordova.plugin.localnotification.ClickReceiver flags=0x0
    time="2022-03-23 13:09:07" type=ACTIVITY_PAUSED package=gov.colorado.energyoffice.emission class=de.appplant.cordova.plugin.localnotification.ClickReceiver instanceId=101733288 taskRootPackage=gov.colorado.energyoffice.emission taskRootClass=de.appplant.cordova.plugin.localnotification.ClickReceiver flags=0x0
    time="2022-03-23 13:09:07" type=ACTIVITY_STOPPED package=gov.colorado.energyoffice.emission class=de.appplant.cordova.plugin.localnotification.ClickReceiver instanceId=101733288 taskRootPackage=gov.colorado.energyoffice.emission taskRootClass=de.appplant.cordova.plugin.localnotification.ClickReceiver flags=0x0
    time="2022-03-23 13:09:08" type=ACTIVITY_RESUMED package=gov.colorado.energyoffice.emission class=de.appplant.cordova.plugin.localnotification.ClickReceiver instanceId=101733288 taskRootPackage=gov.colorado.energyoffice.emission taskRootClass=de.appplant.cordova.plugin.localnotification.ClickReceiver flags=0x0
```

SMS messages app (visible behind) sent to background and then the click receiver also sent to the background.

```
    time="2022-03-23 13:09:08" type=ACTIVITY_PAUSED package=com.motorola.messaging class=com.android.messaging.ui.conversationlist.ConversationListActivity instanceId=1448383 taskRootPackage=com.motorola.messaging taskRootClass=com.android.messaging.ui.conversationlist.ConversationListActivity flags=0x0
    time="2022-03-23 13:09:08" type=KEYGUARD_HIDDEN package=android flags=0x0
    time="2022-03-23 13:09:33" type=SCREEN_NON_INTERACTIVE package=android flags=0x0
    time="2022-03-23 13:09:34" type=ACTIVITY_PAUSED package=gov.colorado.energyoffice.emission class=de.appplant.cordova.plugin.localnotification.ClickReceiver instanceId=101733288 taskRootPackage=gov.colorado.energyoffice.emission taskRootClass=de.appplant.cordova.plugin.localnotification.ClickReceiver flags=0x0
```

>  1:17: click on notification; same hang; emailed screenshot; bar on top goes away


```
    time="2022-03-23 13:17:36" type=USER_INTERACTION package=gov.colorado.energyoffice.emission flags=0x0
    time="2022-03-23 13:17:36" type=ACTIVITY_RESUMED package=gov.colorado.energyoffice.emission class=de.appplant.cordova.plugin.localnotification.ClickReceiver instanceId=178544308 taskRootPackage=gov.colorado.energyoffice.emission taskRootClass=de.appplant.cordova.plugin.localnotification.ClickReceiver flags=0x0
    time="2022-03-23 13:17:36" type=ACTIVITY_PAUSED package=gov.colorado.energyoffice.emission class=de.appplant.cordova.plugin.localnotification.ClickReceiver instanceId=178544308 taskRootPackage=gov.colorado.energyoffice.emission taskRootClass=de.appplant.cordova.plugin.localnotification.ClickReceiver flags=0x0
    time="2022-03-23 13:17:36" type=ACTIVITY_STOPPED package=gov.colorado.energyoffice.emission class=de.appplant.cordova.plugin.localnotification.ClickReceiver instanceId=178544308 taskRootPackage=gov.colorado.energyoffice.emission taskRootClass=de.appplant.cordova.plugin.localnotification.ClickReceiver flags=0x0
    time="2022-03-23 13:17:36" type=ACTIVITY_RESUMED package=gov.colorado.energyoffice.emission class=de.appplant.cordova.plugin.localnotification.ClickReceiver instanceId=178544308 taskRootPackage=gov.colorado.energyoffice.emission taskRootClass=de.appplant.cordova.plugin.localnotification.ClickReceiver flags=0x0
```

> 1:18: launch CanBikeCO manually; launches slowly; notification still present

```
    time="2022-03-23 13:18:56" type=ACTIVITY_RESUMED package=gov.colorado.energyoffice.emission class=gov.colorado.energyoffice.emission.MainActivity instanceId=133170508 taskRootPackage=gov.colorado.energyoffice.emission taskRootClass=gov.colorado.energyoffice.emission.MainActivity flags=0x0
    time="2022-03-23 13:19:39" type=USER_INTERACTION package=gov.colorado.energyoffice.emission flags=0x0
    time="2022-03-23 13:19:39" type=ACTIVITY_PAUSED package=gov.colorado.energyoffice.emission class=gov.colorado.energyoffice.emission.MainActivity instanceId=133170508 taskRoot
Package=gov.colorado.energyoffice.emission taskRootClass=gov.colorado.energyoffice.emission.MainActivity flags=0x0
    time="2022-03-23 13:19:39" type=ACTIVITY_RESUMED package=gov.colorado.energyoffice.emission class=de.appplant.cordova.plugin.localnotification.ClickReceiver instanceId=360049
53 taskRootPackage=gov.colorado.energyoffice.emission taskRootClass=de.appplant.cordova.plugin.localnotification.ClickReceiver flags=0x0
```

**Comment by shankari** [comment]
Can reproduce in the emulator as long as we don't have breakpoints

| Phone app | emulator |
|------------|--------------|
| ![bar_instead_of_opening](https://user-images.githubusercontent.com/2423263/159813802-ec42d4ae-7ba2-4001-9214-6d6b0f8df49d.png) | ![Screenshot_1648078560](https://user-images.githubusercontent.com/2423263/159814246-33977962-9898-4a36-9e2c-d82bf71c4eb3.png) |

Interestingly, when I had breakpoints in the code, it did not hang, so appears to be a timing issue.

**Comment by shankari** [comment]
ok, so we get the `ClickReceiver` notification but don't do anything with it

```
03-23 16:46:12.149  2044  3172 I ActivityTaskManager: START u0 {flg=0x40000000 cmp=edu.berkeley.eecs.emission/de.appplant.cordova.plugin.localnotification.ClickReceiver (has extras)} from uid 10158
03-23 16:46:12.150  2044  3172 W ActivityTaskManager: startActivity called from non-Activity context; forcing Intent.FLAG_ACTIVITY_NEW_TASK for: Intent { flg=0x40000000 cmp=edu.berkeley.eecs.emission/de.appplant.cordova.plugin.localnotification.ClickReceiver (has extras) }
```
On attaching the debugger, the method is not called.

**Comment by shankari** [comment]
So we expect to have `onResume()` invoked on `AbstractClickReceiver` to start this entire process. My guess is that because of this "startActivity called from non-Activity context", we create an activity in a new task (`FLAG_ACTIVITY_NEW_TASK`) instead of associating it with the existing `MainActivity`.

Let's compare with a different permission reset to confirm...

**Comment by shankari** [comment]
Reset the activity permission, but unfortunately saw the same logs

```
03-23 17:02:00.006  2044  2141 I ActivityTaskManager: START u0 {flg=0x40000000 cmp=edu.berkeley.eecs.emission/de.appplant.cordova.plugin.localnotification.ClickReceiver (has extras)} from uid 10158
03-23 17:02:00.007  2044  2141 W ActivityTaskManager: startActivity called from non-Activity context; forcing Intent.FLAG_ACTIVITY_NEW_TASK for: Intent { flg=0x40000000 cmp=edu.berkeley.eecs.emission/de.appplant.cordova.plugin.localnotification.ClickReceiver (has extras) }
```

**Comment by shankari** [comment]
After adding the additional logging, it worked

```
2022-03-23 19:02:04.171 20397-20397/edu.berkeley.eecs.emission I/AbstractClickReceiver: starting onCreate
2022-03-23 19:02:04.191 20397-20397/edu.berkeley.eecs.emission I/AbstractClickReceiver: starting onStart
2022-03-23 19:02:04.192 20397-20397/edu.berkeley.eecs.emission I/AbstractClickReceiver: starting onResume
2022-03-23 19:02:04.192 20397-20397/edu.berkeley.eecs.emission I/AbstractClickReceiver: checking bundle Bundle[mParcelledData.dataSize=280]
2022-03-23 19:02:04.194 20397-20397/edu.berkeley.eecs.emission I/AbstractClickReceiver: found toast {"id":362253744,"title":"Incorrect app settings","text":"Click to view and fix app status","data":{"redirectTo":"root.main.control","redirectParams":{"launchAppStatusModal":true}},"trigger":{"type":"calendar"},"progressBar":{"enabled":false}}
2022-03-23 19:02:04.194 20397-20397/edu.berkeley.eecs.emission I/ClickReceiver: starting onClick
2022-03-23 19:02:04.197 20397-20397/edu.berkeley.eecs.emission I/ClickReceiver: about to launch app if
2022-03-23 19:02:04.197 20397-20397/edu.berkeley.eecs.emission I/ClickReceiver: about to launch app
2022-03-23 19:02:04.201 20397-20397/edu.berkeley.eecs.emission I/ClickReceiver: about to fire event
2022-03-23 19:02:04.202 20397-20397/edu.berkeley.eecs.emission D/CordovaWebViewImpl: >>> loadUrl(javascript:cordova.plugins.notification.local.fireEvent("click",{"id":362253744,"title":"Incorrect app settings","text":"Click to view and fix app status","data":{"redirectTo":"root.main.control","redirectParams":{"launchAppStatusModal":true}},"trigger":{"type":"calendar"},"progressBar":{"enabled":false}},{"event":"click","foreground":false,"queued":false,"notification":362253744}))
2022-03-23 19:02:04.203 20397-20397/edu.berkeley.eecs.emission I/ClickReceiver: about to cancel notification
2022-03-23 19:02:04.209 20397-20397/edu.berkeley.eecs.emission I/AbstractClickReceiver: onClick complete, finishing...
```

**Comment by shankari** [comment]
Ok it failed this time because the toast is null. No changes to the app; so it is an intermittent issue

```
2022-03-23 19:56:05.497 20397-20397/edu.berkeley.eecs.emission I/AbstractClickReceiver: starting onCreate
2022-03-23 19:56:05.505 20397-20397/edu.berkeley.eecs.emission I/AbstractClickReceiver: starting onStart
2022-03-23 19:56:05.506 20397-20397/edu.berkeley.eecs.emission I/AbstractClickReceiver: starting onResume
2022-03-23 19:56:05.506 20397-20397/edu.berkeley.eecs.emission I/AbstractClickReceiver: checking bundle Bundle[mParcelledData.dataSize=280]
2022-03-23 19:56:05.507 20397-20397/edu.berkeley.eecs.emission I/AbstractClickReceiver: found toast null
```

**Comment by shankari** [comment]
so somehow the shared preferences are not really shared. I wonder if this is because they are read as MODE_PRIVATE, and because we don't run the notification code in the UI thread, they are not accessible later.

Let's try moving them into the UI thread first, and experiment with listeners later.

```
    private SharedPreferences getPrefs () {
        return context.getSharedPreferences(PREF_KEY_ID, Context.MODE_PRIVATE);
    }

```

**Comment by shankari** [comment]
After moving it into the UI thread per https://stackoverflow.com/a/64969640/4040267
- 8:57: worked
- 9:58 did not work
- 10:46: worked
- 11:20: did not work
- 11:25: did not work
- 11:35: did not work
- 11:48: did not work

**Comment by shankari** [comment]
For the record, as of right now (Mar 25), the way to see the app reset settings in android 10 or older devices is:
- Play store -> Play Protect -> Settings -> App privacy

**Comment by shankari** [comment]
ANRs are now back to normal.
<img width="1461" alt="Screen Shot 2022-04-05 at 4 42 18 PM" src="https://user-images.githubusercontent.com/2423263/161868603-fc0fa5be-c2d7-4876-966c-f120d57312c8.png">

**Comment by shankari** [comment]
Ran into an ANR again while testing on API 29 emulator and on my API 29 phone.

```
  at java.util.concurrent.locks.LockSupport.park(LockSupport.java:190)
  at androidx.concurrent.futures.AbstractResolvableFuture.get(AbstractResolvableFuture.java:485)
  at edu.berkeley.eecs.emission.cordova.tracker.verification.SensorControlChecks.checkUnusedAppsUnrestricted(SensorControlChecks.java:109)
  at edu.berkeley.eecs.emission.cordova.tracker.verification.SensorControlBackgroundChecker.checkAppState(SensorControlBackgroundChecker.java:150)
  at edu.berkeley.eecs.emission.cordova.tracker.location.TripDiaryStateMachineService.setNewState(TripDiaryStateMachineService.java:121)
  at edu.berkeley.eecs.emission.cordova.tracker.location.TripDiaryStateMachineService$2.lambda$run$0$TripDiaryStateMachineService$2(TripDiaryStateMachineService.java:431)
  at edu.berkeley.eecs.emission.cordova.tracker.location.-$$Lambda$TripDiaryStateMachineService$2$XlU-sYX8UI60FDLW_2Y2NhnV45I.onComplete(lambda:-1)
```

**Comment by shankari** [comment]
Wait, so the version in the CanBikeCO platform is:

```
139           /*
140            * AsyncTask is deprecated, so let's try to use the standard concurrent utilities here.
141            * Haven't used raw java concurrent utilities for 20 years (new Thread(...)) anyone?
142            * So this code snippet is based on
143            * https://stackoverflow.com/a/64969640/4040267
144            *
145            * Do we need to do anything in the UI thread. The `checkUnusedAppsUnrestricted`
146            * doesn't say anything about which thread to call it from. What about
147            * generateOpenAppSettingsNotification? Nothing in the NotificationManager indicates that
148            * it needs to run on the UI thread either.
149            */
150       if (!SensorControlChecks.checkUnusedAppsUnrestricted(ctxt)) {
151             Log.i(ctxt, TAG, "all current settings and permissions are probably valid, but could be reset     later");
152             Log.i(ctxt, TAG, "don't generate a tracking error right now, but let's ask the user to avoid t    he reset ");
153             generateOpenAppSettingsNotification(ctxt);
154       }
155     }
```

what the heck!

**Comment by shankari** [comment]
What is really weird is that the comment above the thread code is retained but not the thread code itself.

<img width="1789" alt="Screen Shot 2022-04-08 at 7 09 30 AM" src="https://user-images.githubusercontent.com/2423263/162452933-a5c79370-2c66-4505-9024-f205f6368429.png">

**Comment by shankari** [comment]
It is even weirder because part of the change did make it in.
<img width="1789" alt="Screen Shot 2022-04-08 at 8 37 47 AM" src="https://user-images.githubusercontent.com/2423263/162475570-a7028fc5-9b33-4ef3-80cf-098643639114.png">

The only explanation I can think of is that I edited the code directly while experimenting with https://github.com/e-mission/e-mission-docs/issues/712

Ok so let's remove and re-add the plugin and generate a new version.

**Comment by shankari** [comment]
That seems to work on my personal phone and in the emulator
Will push to production tonight
